### DISCUSSION TITLE
How to pre-render pages with content from Strapi CMS

### DISCUSSION THREAD

**Discussion by robinkloeckner** [nuxt/nuxt#24172] [Questions]
In my Nuxt 3 project, I want to pre-render several pages during build time. The content of these pages reside in a Strapi v4 CMS for which I use the Strapi Nuxt module to fetch the api.

In my `nuxt.config.ts` is set the corresponding routes as follows to pre-render them during build time ([Nuxt documentation](https://nuxt.com/docs/guide/concepts/rendering#hybrid-rendering))
```
export default defineNuxtConfig({
  routeRules: {
    "/posts/**": { prerender: true },
  }
});
```

In my component I wrap the `find` function with `useAsyncData()` since I want to prevent the content to be fetched again client-side. Subsequently I want to use the fetched content to be rendered in the HTML template and I also want to use the content in the setup function for different tasks as well - for example to manipulate the meta tags. To prevent this to happen client-side I execute the corresponding function only if `process.server` is true.

```
const {data}  = await useAsyncData(
    "content",
    async () => {
      const response = await find(
          "url",
          {
            ...
          });

      return response.data;
    }
);

if (process.server) {
  functionToManipulateMetaTags(data.value.attributes);
}

...
```

After `nuxi build` completes the app is deployed on a VPS. In the browser, the page looks as expected, but on every page load, the content is fetched from the CMS but I expect the page to be pre-rendered during build time. Also using something like `if (process.server) {...}` to prevent executing functions client-side that are meant to process the fetched content during build or at least server-side does not seem to be the correct approach. 

Can somebody point me to the right direction?

**Comment by robinkloeckner** [comment]
After adding the routes to pre-render manually to the `nitro:config` hook in the `nuxt.config.ts` (following #13949 ), the files of the corresponding routes can be found in the `.output` directory after the build process. However after deploying the application and navigating to a pre-rendered route by a hyperlink for which I use a NuxtLink component, the api is still fetched client-side. Since this is not the case when accessing the pre-rendered route directly by entering the URL in the browser address bar I expect the problem isn't related to Strapi.

**Comment by renardsas** [comment]
> After adding the routes to pre-render manually to the `nitro:config` hook in the `nuxt.config.ts` (following #13949 ), the files of the corresponding routes can be found in the `.output` directory after the build process. However after deploying the application and navigating to a pre-rendered route by a hyperlink for which I use a NuxtLink component, the api is still fetched client-side. Since this is not the case when accessing the pre-rendered route directly by entering the URL in the browser address bar I expect the problem isn't related to Strapi.

If I understand correctly what you're looking for, try this new parameter (if you're using > 3.8): getCachedData.

[https://nuxt.com/blog/v3-8#️-data-fetching-improvements](https://nuxt.com/blog/v3-8#%EF%B8%8F-data-fetching-improvements)
https://github.com/nuxt/nuxt/pull/20747

**Comment by robinkloeckner** [comment]
@renardsas thanks. I had the same in mind, but tried it with Nuxt v3.8.2 but having the same behavior. Every time in navigate to a static route via a link created with NuxtLink, content from the API is fetched again.

**Reply by WaldemarEnns** [reply]
Any solution found for this?

**Comment by robinkloeckner** [comment]
@WaldemarEnns Unfortunately, I could not pinpoint the issue, so I ended up rebuilding the entire project

**Reply by WaldemarEnns** [reply]
Have you ever tried out this?
https://nuxt.com/docs/getting-started/prerendering#prerenderroutes-nuxt-hook

I'll try it out now.

**Comment by robinkloeckner** [comment]
I use the `nitro:config` hook to push routes to `prerender.routes` in the nitro config described [here under Method 1](https://www.docs4.dev/posts/nuxt-3-hybrid-rendering-pre-render-dynamic-routes-ssg) and here [#13949](https://github.com/nuxt/nuxt/issues/13949). But on a first glance it could be the same hook?

I implemented this method before re-buildung the project. So in my case this wasn't the issue. The pages were pre-rendered but still re-fetched the contents from Strapi client-side.

**Comment by lcberg** [comment]
Did you try the vue ssr onServerPrefetch hook? [Docs](https://vuejs.org/api/composition-api-lifecycle.html#onserverprefetch)
